File: modules/react-native-nitro-refresh/ios/HybridRefreshController.swift
import Foundation
import NitroModules

/// Fabric 刷新视图暴露给 Nitro 控制器的最小命令接口。
/// 使用 class-only 协议以便注册表安全地持有弱引用。
@objc public protocol NitroRefreshViewBinding: AnyObject {
  func setRefreshingFromController(_ refreshing: Bool)
  func beginRefreshFromController()
  func cancelRefreshFromController()
  func finishRefreshFromController(_ result: String, resultDuration: Double)
  func pullToMaxFromController()
}

private final class WeakController {
  weak var value: HybridRefreshController?

  init(_ value: HybridRefreshController?) {
    self.value = value
  }
}

private final class WeakBinding {
  weak var value: (any NitroRefreshViewBinding)?

  init(_ value: (any NitroRefreshViewBinding)?) {
    self.value = value
  }
}

/// 连接 Nitro HybridObject 与 Fabric ComponentView 的进程内注册表。
@objc(NitroRefreshControllerRegistry)
public final class NitroRefreshControllerRegistry: NSObject {
  private static var controllers: [String: WeakController] = [:]
  private static var bindings: [String: WeakBinding] = [:]
  private static let lock = NSLock()

  fileprivate static func register(_ controller: HybridRefreshController) {
    lock.lock()
    controllers[controller.id] = WeakController(controller)
    let binding = bindings[controller.id]?.value
    lock.unlock()

    if let binding {
      controller.attach(binding)
    }
  }

  fileprivate static func unregister(_ controller: HybridRefreshController) {
    lock.lock()
    if controllers[controller.id]?.value === controller {
      controllers.removeValue(forKey: controller.id)
    }
    lock.unlock()
  }

  @objc(attachControllerId:binding:)
  public static func attach(controllerId: String, binding: any NitroRefreshViewBinding) {
    lock.lock()
    bindings[controllerId] = WeakBinding(binding)
    let controller = controllers[controllerId]?.value
    lock.unlock()
    controller?.attach(binding)
  }

  @objc(detachControllerId:binding:)
  public static func detach(controllerId: String, binding: any NitroRefreshViewBinding) {
    lock.lock()
    if bindings[controllerId]?.value === binding {
      bindings.removeValue(forKey: controllerId)
    }
    let controller = controllers[controllerId]?.value
    lock.unlock()
    controller?.detach(binding)
  }

  @objc(requestRefreshForControllerId:)
  public static func requestRefresh(controllerId: String) {
    controller(for: controllerId)?.requestRefresh()
  }

  @objc(notifyControllerId:phase:)
  public static func notify(controllerId: String, phase: String) {
    guard let phase = RefreshPhase(fromString: phase) else { return }
    controller(for: controllerId)?.notifyPhase(phase)
  }

  @objc(updateControllerId:phase:offset:refreshing:)
  public static func update(
    controllerId: String,
    phase: String,
    offset: Double,
    refreshing: Bool
  ) {
    guard let phase = RefreshPhase(fromString: phase) else { return }
    controller(for: controllerId)?.updateState(
      phase: phase,
      offset: offset,
      refreshing: refreshing
    )
  }

  private static func controller(for id: String) -> HybridRefreshController? {
    lock.lock()
    let controller = controllers[id]?.value
    if controller == nil {
      controllers.removeValue(forKey: id)
    }
    lock.unlock()
    return controller
  }
}

/// Nitro RefreshController 的 iOS 实现。
final class HybridRefreshController: HybridRefreshControllerSpec {
  let id = UUID().uuidString

  private var onRefresh: (() -> Void)?
  private var onStateChange: ((RefreshPhase) -> Void)?
  private weak var binding: (any NitroRefreshViewBinding)?
  private var requestedRefreshing = false
  private let stateLock = NSLock()
  private var latestPhase: RefreshPhase = .idle
  private var latestOffset = 0.0
  private var latestRefreshing = false

  override init() {
    super.init()
    NitroRefreshControllerRegistry.register(self)
  }

  func setOnRefresh(callback: @escaping () -> Void) throws {
    NitroRefreshControllerRegistry.register(self)
    onRefresh = callback
  }

  func setOnStateChange(callback: @escaping (RefreshPhase) -> Void) throws {
    onStateChange = callback
  }

  func clearCallbacks() throws {
    onRefresh = nil
    onStateChange = nil
    NitroRefreshControllerRegistry.unregister(self)
  }

  func beginRefresh() throws {
    DispatchQueue.main.async { [weak self] in
      self?.binding?.beginRefreshFromController()
    }
  }

  func cancelRefresh() throws {
    requestedRefreshing = false
    DispatchQueue.main.async { [weak self] in
      self?.binding?.cancelRefreshFromController()
    }
  }

  func finishRefresh(refreshResult: RefreshResult, resultDuration: Double) throws {
    requestedRefreshing = false
    DispatchQueue.main.async { [weak self] in
      self?.binding?.finishRefreshFromController(
        refreshResult.stringValue,
        resultDuration: resultDuration
      )
    }
  }

  func getState() throws -> RefreshStateSnapshot {
    stateLock.lock()
    let phase = latestPhase
    let offset = latestOffset
    let refreshing = latestRefreshing
    stateLock.unlock()
    return RefreshStateSnapshot(
      phase: phase,
      offset: offset,
      refreshing: refreshing
    )
  }

  func pullToMax() throws {
    DispatchQueue.main.async { [weak self] in
      self?.binding?.pullToMaxFromController()
    }
  }

  func setRefreshing(refreshing: Bool) throws {
    requestedRefreshing = refreshing
    DispatchQueue.main.async { [weak self] in
      self?.binding?.setRefreshingFromController(refreshing)
    }
  }

  fileprivate func attach(_ binding: any NitroRefreshViewBinding) {
    self.binding = binding
    DispatchQueue.main.async { [weak self, weak binding] in
      guard let self, let binding else { return }
      binding.setRefreshingFromController(self.requestedRefreshing)
    }
  }

  fileprivate func detach(_ binding: any NitroRefreshViewBinding) {
    if self.binding === binding {
      self.binding = nil
    }
  }

  fileprivate func requestRefresh() {
    requestedRefreshing = true
    onRefresh?()
  }

  fileprivate func notifyPhase(_ phase: RefreshPhase) {
    onStateChange?(phase)
  }

  fileprivate func updateState(
    phase: RefreshPhase,
    offset: Double,
    refreshing: Bool
  ) {
    stateLock.lock()
    latestPhase = phase
    latestOffset = offset
    latestRefreshing = refreshing
    stateLock.unlock()
  }
}

File: modules/react-native-nitro-refresh/android/src/main/java/com/margelo/nitro/refresh/HybridRefreshController.kt
package com.margelo.nitro.refresh

import java.lang.ref.WeakReference
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap

/**
 * Nitro HybridObject 的 Android 实现。
 *
 * 该对象只保存 React 的受控刷新意图和离散回调；实际手势、连续位移与动画均由
 * [NitroRefreshLayout] 处理，避免在高频更新路径中维护 JS 不会读取的状态副本。
 */
class HybridRefreshController : HybridRefreshControllerSpec() {
  override val id: String = UUID.randomUUID().toString()

  private var onRefresh: (() -> Unit)? = null
  private var onStateChange: ((RefreshPhase) -> Unit)? = null
  @Volatile private var requestedRefreshing = false
  private var binding = WeakReference<NitroRefreshLayout>(null)

  init {
    // Fabric 视图可能晚于控制器挂载，先登记以支持视图随后通过 id 查找。
    controllers[id] = WeakReference(this)
  }

  override fun setOnRefresh(callback: () -> Unit) {
    // React Strict Mode 会执行 setup -> cleanup -> setup，重新登记可支持第二次 setup。
    controllers[id] = WeakReference(this)
    onRefresh = callback
  }

  override fun setOnStateChange(callback: (phase: RefreshPhase) -> Unit) {
    onStateChange = callback
  }

  override fun clearCallbacks() {
    onRefresh = null
    onStateChange = null
    controllers.computeIfPresent(id) { _, reference ->
      if (reference.get() === this) null else reference
    }
  }

  override fun setRefreshing(refreshing: Boolean) {
    // 即使视图尚未挂载也保存意图，attach 时会立即补同步初始 refreshing=true。
    requestedRefreshing = refreshing
    binding.get()?.setRefreshingFromController(refreshing)
  }

  internal fun attach(view: NitroRefreshLayout) {
    binding = WeakReference(view)
    view.setRefreshingFromController(requestedRefreshing)
  }

  internal fun detach(view: NitroRefreshLayout) {
    if (binding.get() === view) {
      binding.clear()
    }
  }

  internal fun requestRefresh() {
    // 用户手势先进入 refreshing，再通知 JS 回写受控属性。
    requestedRefreshing = true
    onRefresh?.invoke()
  }

  internal fun notifyPhase(phase: RefreshPhase) {
    onStateChange?.invoke(phase)
  }

  companion object {
    /** 所有控制器的进程内弱引用索引，键与 Fabric 的 controllerId 属性一致。 */
    private val controllers = ConcurrentHashMap<String, WeakReference<HybridRefreshController>>()

    internal fun find(id: String): HybridRefreshController? {
      val controller = controllers[id]?.get()
      if (controller == null) {
        controllers.remove(id)
      }
      return controller
    }
  }
}
